Design proposal for delayed certificate activation (#7289)
cert-manager/cert-manager: design/20240918.delayed-certificate-activation.md

## Proposal

Introduce a new optional field, `NotBefore`, to the spec of both `Certificate` and `CertificateRequest` resources. This field will allow users to explicitly define the start time of the certificate's validity period.

### API changes

```
type CertificateSpec struct {
  // EXISTING FIELDS
  // ...

  // NotBefore defines the time when the certificate becomes valid.
  // If omitted or set to a time in the past, the certificate will be 
  // valid immediately upon issuance.
  // +optional
  NotBefore *metav1.Time `json:"notBefore,omitempty"`
}

type CertificateRequestSpec struct {
  // EXISTING FIELDS
  // ...

  // NotBefore defines the time when the certificate becomes valid.
  // If omitted or set to a time in the past, the certificate will be 
  // valid immediately upon issuance.
  // +optional
  NotBefore *metav1.Time `json:"notBefore,omitempty"` 
}
```

### Examples

#### Issuance with Future NotBefore

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-cert
spec:
  secretName: example-cert-tls
  issuerRef:
    name: letsencrypt-prod
    kind: ClusterIssuer
  commonName: example.com

>   notBefore: "2024-10-01T10:00:00Z" 

Reviewer: suggestion: I'm not aware of anywhere in the cert-manager codebase where we currently have a need to parse user-specified dates like this. That's quite a big thing to add to our API since there are multiple complications:

1. Time zones (which need a time zone database to be able to be parsed accurately)
2. Date formats (of which there are many).

I think we'd need to address that issue in this design document if we want users to be able to specify exact dates in a string format. We _could_ say "all dates must be specified in RFC3339 format" which is a good start.

Author: Added a comment in API definition.

  duration: 2160h # 90 days
```

This example creates a certificate that will only become valid on October 1st, 2024, even if it is issued earlier.

#### Renewal with NotBefore

> During renewal, cert-manager will use the `NotBefore` value from the `Certificate` or `CertificateRequest` CR if it's still in the future. Otherwise, the renewed certificate will be issued with immediate validity.

Reviewer 2: ```suggestion
During renewal, cert-manager will use the `NotBefore` value from the `Certificate` or `CertificateRequest` CR if it's still in the future. Otherwise, the renewed certificate will be issued with `NotBefore` set to now.
```

Author: Done.

> When a certificate with a future `NotBefore` is renewed, cert-manager should preserve the original `NotBefore` time if it's still in the future. If the original `NotBefore` is in the past, the renewed certificate should be valid immediately.

Reviewer 2: What should the value of `NotBefore` be if the certificate's `NotBefore` was in the past?

Author: certificate's `NotBefore` value will depend on the value of `NotBefore` field set in `Certificate` or `CertificateRequest` CR. If the `NotBefore` in spec is set as nil, certificate's `NotBefore` will be set as `time.Now()`.  If the `NotBefore` in spec is set as future time, certificate's `NotBefore` will be set as `spec.NotBefore`. If the `NotBefore` in spec is set as past time, certificate's `NotBefore` will be set as `time.Now()`.

From the implementation perspective, [this](https://github.com/jaingeet/cert-manager/blob/master/pkg/util/pki/certificatetemplate.go#L56-L62) function will be modified as below  - 
```
func CertificateTemplateOverrideDuration(duration time.Duration, notBefore *metav1.Time) CertificateTemplateValidatorMutator {
	return func(req *x509.CertificateRequest, cert *x509.Certificate) error {
		if notBefore != nil && notBefore.Time.After(time.Now()) {
			cert.NotBefore = notBefore.Time
		} else {
			cert.NotBefore = time.Now()
		}

		cert.NotAfter = cert.NotBefore.Add(duration)
		return nil
	}
}
```

Reviewer 2: That's a reasonable behavior, but not obvious from line 115. Could you adjust:

> If the original `NotBefore` is in the past

to clarify this bit?

Reviewer 2: Note that you really shouldn't call `time.Now()` twice like that. It should be called once:
```go
		now := time.Now()
		if notBefore != nil && notBefore.Time.After(now) {
			cert.NotBefore = notBefore.Time
		} else {
			cert.NotBefore = now
		}
```

Author: Updated the doc with the suggestion.

Reviewer 2: I think I'd change `valid immediately` to `set to now`.

---
title: Delayed Certificate Activation
authors:
  - "@jaingeet"
reviewers:
 - @jetstack/team-cert-manager
approvers:
 - @jetstack/team-cert-manager
editor: "@jaingeet"
creation-date: 2024-09-18
last-updated: 2024-09-18
status: implementable
---

# Delayed Certificate Activation

## Table of Contents

<!-- toc -->
- [Delayed Certificate Activation](#delayed-certificate-activation)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
  - [Proposal](#proposal)
    - [API changes](#api-changes)
    - [Examples](#examples)
      - [Issuance with Future NotBefore](#issuance-with-future-notbefore)
      - [Renewal with NotBefore](#renewal-with-notbefore)
    - [Implementation Considerations](#implementation-considerations)
    - [Backward Compatibilty](#backward-compatibilty)
    - [Test Plan](#test-plan)
    - [Upgrading](#upgrading)
    - [Misc](#misc)
  - [Alternatives](#alternatives)
<!-- /toc -->

## Summary

> This proposal introduces a mechanism to specify a future `NotBefore` time in cert-manager certificates, enabling delayed activation of issued certificates.

Reviewer 2: ```suggestion
This proposal introduces a mechanism to specify a future `NotBefore` time in cert-manager issued certificates, enabling delayed activation of issued certificates.
```

## Motivation

Currently, cert-manager creates the Certificate Signing Request (CSR) with `NotBefore` set to the current time (`time.Now()`).  This prevents use cases where a certificate needs to be issued in advance but only become valid at a specific future time. For example,

- **Pre-provisioning Certificates:** Generating certificates for resources that will be deployed in the future, such as for a scheduled application rollout.
- **Compliance Requirements:** Adhering to specific security policies that mandate a future validation start date for certificates.

> 

Reviewer 2: ```suggestion
- **Synchronization Requirements:** Partners require certificates to be shared in advance so they can update systems to trust them before they're deployed.

```

### API changes

```
type CertificateSpec struct {
  // EXISTING FIELDS
  // ...

  // NotBefore defines the time when the certificate becomes valid.
  // If omitted or set to a time in the past, the certificate will be 
  // valid immediately upon issuance.
  // +optional
  NotBefore *metav1.Time `json:"notBefore,omitempty"`
}

type CertificateRequestSpec struct {
  // EXISTING FIELDS
  // ...

  // NotBefore defines the time when the certificate becomes valid.
  // If omitted or set to a time in the past, the certificate will be 
  // valid immediately upon issuance.
  // Date must be specified in RFC3339 format.
  // +optional
  NotBefore *metav1.Time `json:"notBefore,omitempty"` 
}
```

### Examples

#### Issuance with Future NotBefore

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-cert
spec:
  secretName: example-cert-tls
  issuerRef:
    name: letsencrypt-prod
    kind: ClusterIssuer
  commonName: example.com
  notBefore: "2024-10-01T10:00:00Z" 
  duration: 2160h # 90 days
```

> This example creates a certificate that will only become valid on October 1st, 2024, even if it is issued earlier.

Reviewer 2: This example is in the past. It might make sense to push the example date a couple of decades into the future, or to change the end of the prose `even if it is issued earlier` to `even if it was requested on January 1st, 2023`.
```suggestion
This example creates a certificate that would only become valid on October 1st, 2024, even if it was requested on January 1st, 2023.
```

#### Renewal with NotBefore

During renewal, cert-manager will use the `NotBefore` value from the `Certificate` or `CertificateRequest` CR if it's still in the future. Otherwise, the renewed certificate will be issued with `NotBefore` set to the current time.

### Implementation Considerations

**Defaults:** If `NotBefore` is omitted or set to a time in the past (this can happen during renewal), the current time should be used.

### Backward Compatibilty

This proposal maintains backward compatibility. Existing `Certificate` and `CertificateRequest` resources without the new field will continue to function as they currently do, with `NotBefore` defaulting to the current time.

### Test Plan

> **Unit Tests:** Comprehensive unit tests will be added to cover different `NotBefore` values, including nil, past, and future times.

Reviewer 2: ```suggestion
**Unit Tests:** Comprehensive unit tests will be added to cover different `NotBefore` values, including nil, past, and future times as well as invalid inputs.
```

**Integration Tests:** End-to-end tests will be added to verify the complete certificate issuance flow with various `NotBefore` values.

### Upgrading

As this is an optional field, upgrading to a cert-manager version with this new field or downgrading to one without it should not require extra steps or cause unnecessary re-issuances.

### Misc

> The cert-manager documentation will be updated to include details about this feature, usage with examples.

Reviewer 2: ```suggestion
The cert-manager documentation will be updated to include details about using this feature with examples.
```
